binding.c

```diff
@@ -568,81 +666,100 @@ on_udx_stream_firewall (udx_stream_t *stream, udx_socket_t *socket, const struct
+  js_value_t *ctx;
+  err = js_get_reference_value(env, n->ctx, &ctx);
+  assert(err == 0);
 
-  napi_value callback;
-  napi_get_reference_value(env, n->on_firewall, &callback);
+  js_value_t *callback;
+  err = js_get_reference_value(env, n->on_firewall, &callback);
+  assert(err == 0);
 
-  napi_value res;
-  napi_value argv[4];
+  js_value_t *res;
+  js_value_t *argv[4];
 
-  napi_get_reference_value(env, s->ctx, &(argv[0]));
-  napi_create_uint32(env, port, &(argv[1]));
-  napi_create_string_utf8(env, ip, NAPI_AUTO_LENGTH, &(argv[2]));
-  napi_create_uint32(env, family, &(argv[3]));
+  err = js_get_reference_value(env, s->ctx, &(argv[0]));
+  assert(err == 0);
+  err = js_create_uint32(env, port, &(argv[1]));
+  assert(err == 0);
+  err = js_create_string_utf8(env, (utf8_t *) ip, -1, &(argv[2]));
+  assert(err == 0);
+  err = js_create_uint32(env, family, &(argv[3]));
+  assert(err == 0);
 
-  if (napi_make_callback(env, NULL, ctx, callback, 4, argv, &res) == napi_pending_exception) {
-    napi_value fatal_exception;
-    napi_get_and_clear_last_exception(env, &fatal_exception);
-    napi_fatal_exception(env, fatal_exception);
+  err = js_call_function_with_checkpoint(env, ctx, callback, 4, argv, &res);
+  if (err == 0) {
+    err = js_get_value_uint32(env, res, &fw);
+    assert(err == 0);
   } else {
-    napi_get_value_uint32(env, res, &fw);
+    clear_pending(env, err);
   }
 
-  napi_close_handle_scope(env, scope);
+  err = js_close_handle_scope(env, scope);
+  assert(err == 0);
 
   return fw;
 }
 
 static void
 on_udx_stream_remote_changed (udx_stream_t *stream) {
+  int err;
   udx_napi_stream_t *n = (udx_napi_stream_t *) stream;
   if (n->udx->exiting) return;
 
-  napi_env env = n->env;
+  js_env_t *env = n->env;
 
-  napi_handle_scope scope;
-  napi_open_handle_scope(env, &scope);
+  js_handle_scope_t *scope;
+  err = js_open_handle_scope(env, &scope);
+  assert(err == 0);
 
-  napi_value ctx;
-  napi_get_reference_value(env, n->ctx, &ctx);
+  js_value_t *ctx;
+  err = js_get_reference_value(env, n->ctx, &ctx);
+  assert(err == 0);
 
-  napi_value callback;
-  napi_get_reference_value(env, n->on_remote_changed, &callback);
+  js_value_t *callback;
+  err = js_get_reference_value(env, n->on_remote_changed, &callback);
+  assert(err == 0);
 
-  if (napi_make_callback(env, NULL, ctx, callback, 0, NULL, NULL) == napi_pending_exception) {
-    napi_value fatal_exception;
-    napi_get_and_clear_last_exception(env, &fatal_exception);
-    napi_fatal_exception(env, fatal_exception);
-  }
+  // TODO: tricky!
+  // Due branch linked below
+  // https://github.com/holepunchto/libudx/blob/de408f4ae9cde4139938b09b4f443f2d45f22c9b/src/udx.c#L2577-L2580
+  // this function is either called on IO / udx:process_packet()
+  // or immediately on parent call "udx_napi_stream_change_remote" which is on current js-context.
+  err = js_call_function_with_checkpoint(env, ctx, callback, 0, NULL, NULL);
+  clear_pending(env, err);
```

Comment:
@kasperisager what is the best course of action here? 
tl;dr `on_udx_stream_remote_changed` can be called both by JS and I/O.

Reply:
@mafintosh Anything we can do on the `libudx` side to avoid this? Having a callback where we can't reason about the state of the JavaScript execution stack makes Node-API compatibility difficult.

Reply:
fixed via https://github.com/holepunchto/libudx/pull/243  +  [6f61b85](https://github.com/holepunchto/udx-native/pull/59/commits/6f61b85cf0ad385343453566c9564f73093f22e0) 

Now libUDX `udx_stream_change_remote` only every calls the callback from I/O.

If change is possible immediately the udx call returns 1;
In which case we invoke the remote-changed callback without a checkpoint and then return.